### PR TITLE
:sparkles: [Reset unsafe password] Force reset password on app

### PR DESCRIPTION
**Pull request by tcheymol**

### CODE REVIEW

File: src/ManagerV2/UserManager.php [hunk 1]
```diff
@@ -136,4 +142,26 @@ private function getNextHomonymUsername(string $baseUsername, array $homonyms):
...
+            if (!$authCode) {
+                $this->codeGenerator->generateAndSend($user);
+                $user->setMfaPending(true);
+            } else {
+                $user->setMfaValid($authCode === $user->getEmailAuthCode());
+            }
+        }
+
+        $this->em->flush();
+    }
```
Comment (AntoineAresu) [inline src/ManagerV2/UserManager.php:159]: I wonder if this should live in Domain namespace ?

File: src/EventSubscriber/Api/Oauth2UserResolveSubscriber.php [hunk 2]
```diff
@@ -24,6 +28,11 @@ public function checkUserPasswordValidity(UserResolveEvent $event): UserResolveE
         $isUserPasswordInvalid = $user instanceof LegacyPasswordAuthenticatedUserInterface && !$this->hasher->isPasswordValid($user, $event->getPassword());
         $event->setUser($isUserPasswordInvalid ? null : $user);
 
+        if (!$user->hasPasswordWithLatestPolicy() && $this->passwordHelper->isStrongPassword($event->getPassword())) {
+            $user->setHasPasswordWithLatestPolicy(true);
+            $this->em->flush();
+        }
+
         return $event;
```
Comment (NilsRec) [inline src/EventSubscriber/Api/Oauth2UserResolveSubscriber.php:36]: I think we also have this code in the authenticator class. Maybe we could use a service or something ?

Comment (NilsRec) [reply src/EventSubscriber/Api/Oauth2UserResolveSubscriber.php:36]: Like `AuthenticationProcess` and centralize all checks we do on the user during authentication on web or api

File: src/ManagerV2/UserManager.php [hunk 3]
```diff
@@ -136,4 +142,26 @@ private function getNextHomonymUsername(string $baseUsername, array $homonyms):
     {
         return sprintf('%s-%d', $baseUsername, $this->getMaxHomonymIndex($baseUsername, $homonyms) + 1);
     }
+
+    public function checkMfaCode(User $user, ?string $authCode): void
```
Comment (NilsRec) [inline src/ManagerV2/UserManager.php:146]: `checkAuthCode` as the prop is named authCode ?

File: src/ManagerV2/UserManager.php [hunk 4]
```diff
@@ -136,4 +142,26 @@ private function getNextHomonymUsername(string $baseUsername, array $homonyms):
...
+    {
+        $isMfaEnabled = $user->isMfaEnabled() && $this->appli2faEnabled;
+        if ($isMfaEnabled) {
+            if (!$authCode) {
```
Comment (NilsRec) [inline src/ManagerV2/UserManager.php:150]:
```
if (!$isMfaEnabled) {
    return;
}
```
And maybe the var is useless 
`if (!$user->isMfaEnabled || !$this->appli2faEnabled) `

File: src/ManagerV2/UserManager.php [hunk 5]
```diff
@@ -136,4 +142,26 @@ private function getNextHomonymUsername(string $baseUsername, array $homonyms):
...
+                $user->setMfaPending(true);
+            } else {
+                $user->setMfaValid($authCode === $user->getEmailAuthCode());
+            }
+        }
+
+        $this->em->flush();
+    }
+
+    public function checkUpdateWeakPassword(User $user, mixed $newPassword): void
```
Comment (NilsRec) [inline src/ManagerV2/UserManager.php:161]: Not sure about this method name

File: src/Domain/PasswordStrength/WeakPasswordUpgrader.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+        if (!$user->hasPasswordWithLatestPolicy() && $this->passwordHelper->isStrongPassword($password)) {
+            $user->setHasPasswordWithLatestPolicy(true);
+            $this->em->flush();
+        }
+    }
+
+    public function isDue(User $user, string $password): void
+    {
+        !$user->hasPasswordWithLatestPolicy() && $this->passwordHelper->isStrongPassword($password);
+    }
```
Comment (AntoineAresu) [inline src/Domain/PasswordStrength/WeakPasswordUpgrader.php:30]: You refacto Authenticator code and implement this method

**Review by AntoineAresu — APPROVED** [review]

**Pull request merged**